File: hello-websocket-dart/server/ws_server.dart
// Hello WebSocket Protocol - Dart Server
import 'dart:async';
import 'dart:io';
import 'dart:typed_data';

import '../common/codec.dart';

void main() async {
  final portEnv = Platform.environment['WS_PORT'];
  final p = portEnv != null ? int.tryParse(portEnv) ?? port : port;

  log('ws-server', 'Starting Dart WebSocket server on port $p');

  final server = await HttpServer.bind(InternetAddress.anyIPv4, p);
  await for (final request in server) {
    if (WebSocketTransformer.isUpgradeRequest(request)) {
      final userId = request.headers.value('userId') ??
          'dart-${DateTime.now().millisecondsSinceEpoch}';
      final ws = await WebSocketTransformer.upgrade(request);
      _handleConnection(ws, userId);
    } else {
      request.response.statusCode = HttpStatus.badRequest;
      await request.response.close();
    }
  }
}

void _handleConnection(WebSocket ws, String userId) {
  final sessionId = '${DateTime.now().millisecondsSinceEpoch}';
  var clientLanguage = 'unknown';
  var lastPongTs = nowMs();

  log('ws-server', '[$userId] session+');

  // Background: PING every 1s
  Timer.periodic(Duration(milliseconds: pingIntervalMs), (t) {
    if (ws.readyState != WebSocket.open) {
      t.cancel();
      return;
    }
    ws.add(ping(nowMs()).encode());
  });

  // Background: TIME_NOTIFICATION every 5s
  Timer.periodic(Duration(milliseconds: timeIntervalMs), (t) {
    if (ws.readyState != WebSocket.open) {
      t.cancel();
      return;
    }
    ws.add(timeNotif(nowMs(), nowISO()).encode());
  });

  // Background: KISS_REQUEST every 5s
  Timer.periodic(Duration(milliseconds: kissIntervalMs), (t) {
    if (ws.readyState != WebSocket.open) {
      t.cancel();
      return;
    }
    ws.add(kissRequest(
            Platform.operatingSystem, 'unknown', 'unknown', Platform.version)
        .encode());
  });

  // Background: timeout check every 5s
  Timer.periodic(Duration(seconds: 5), (t) {
    if (ws.readyState != WebSocket.open) {
      t.cancel();
      return;
    }
    if (nowMs() - lastPongTs > sessionTimeoutMs) {
      log('ws-server', '[$userId] session timeout');
      ws.close();
      t.cancel();
    }
  });

  // Receive loop
  ws.listen(
    (data) {
      if (data is! List<int>) return;
      if (data.length > 1024 * 1024) {
        ws.close(WebSocketStatus.messageTooBig, 'message exceeds 1 MiB');
        return;
      }
      final bytes = Uint8List.fromList(data);
      Message msg;
      try {
        msg = decodeMessage(bytes);
      } catch (e) {
        log('ws-server', 'Decode error: $e');
        final unknown = e.toString().contains('unknown message type');
        ws.add(errorMsg(unknown ? errUnknownMsgType : errDecode, e.toString())
            .encode());
        if (!unknown)
          ws.close(WebSocketStatus.protocolError, 'invalid protocol frame');
        return;
      }

      switch (msg.type) {
        case msgHello:
          clientLanguage = msg.clientLanguage!;
          log('ws-server',
              'HELLO from $clientLanguage, session=$sessionId, time=${nowMs()}');
          ws.add(bonjour(serverLang).encode());
          break;

        case msgEchoRequest:
          log('ws-server',
              'ECHO_REQUEST id=${msg.echoId} meta=${msg.echoMeta} data=${msg.echoData}');
          final resp = Message(msgEchoResponse)
            ..echoStatus = 200
            ..echoResults = [
              EchoResult(nowMs(), 0, {
                'id': msg.echoId.toString(),
                'idx': msg.echoData!,
                'data': msg.echoData!,
                'meta': clientLanguage,
              })
            ];
          ws.add(resp.encode());
          break;

        case msgKissResponse:
          log('ws-server',
              'KISS_RESPONSE lang=${msg.kissLanguage} enc=${msg.kissEncoding} tz=${msg.kissTimeZone}');
          break;

        case msgPong:
          lastPongTs = nowMs();
          log('ws-server', 'PONG ts=${msg.timestampMs}');
          break;

        case msgRandomNumber:
          log('ws-server',
              'RANDOM_NUMBER id=${msg.randomId} number=${msg.randomNumber}');
          final hash = hashNumber(msg.randomNumber!);
          ws.add(hashResponseMsg(msg.randomId!, hash).encode());
          log('ws-server', 'HASH_RESPONSE id=${msg.randomId} hash=$hash');
          break;

        case msgDisconnect:
          log('ws-server', 'DISCONNECT reason=${msg.disconnectReason}');
          ws.close();
          break;

        case msgError:
          log('ws-server',
              'ERROR code=${msg.errorCode} msg=${msg.errorMessage}');
          break;

        default:
          log('ws-server',
              'Unknown message type: 0x${msg.type.toRadixString(16)}');
          ws.add(errorMsg(errUnknownMsgType,
                  'unknown type 0x${msg.type.toRadixString(16)}')
              .encode());
      }
    },
    onDone: () {
      log('ws-server', '[$userId] session-');
    },
    onError: (e) {
      log('ws-server', 'Error: $e');
    },
  );
}

File: hello-websocket-kotlin/server/src/main/kotlin/org/feuyeux/ws/server/WsServer.kt
// Hello WebSocket Protocol - Kotlin Server
// Implements the full PROTOCOL.md server lifecycle: handshake, background tasks,
// echo, kiss, ping/pong, time broadcast, random/hash, disconnect.
package org.feuyeux.ws.server

import io.ktor.server.application.*
import io.ktor.server.engine.*
import io.ktor.server.netty.*
import io.ktor.server.routing.*
import io.ktor.server.websocket.*
import io.ktor.websocket.*
import kotlinx.coroutines.*
import org.feuyeux.ws.common.*
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.atomic.AtomicLong

fun main() {
    val port = System.getenv("WS_PORT")?.toIntOrNull() ?: PORT

    log("ws-server", "Starting Kotlin WebSocket server on port $port")

    embeddedServer(Netty, port, host = "0.0.0.0") {
        install(WebSockets)
        routing {
            webSocket("/") {
                handleClient(this)
            }
        }
    }.start(wait = true)
}

suspend fun handleClient(ws: WebSocketServerSession) = coroutineScope {
    val userId = ws.call.request.headers["userId"] ?: "kotlin-${nowMs()}"
    val sessionId = nowMs().toString()
    var clientLanguage = "unknown"
    val lastPongTs = AtomicLong(nowMs())
    val active = AtomicBoolean(true)

    log("ws-server", "[$userId] session+")

    // Background: PING every 1s
    val pingJob = launch {
        while (active.get()) {
            delay(PING_INTERVAL_MS)
            if (!active.get()) break
            try {
                ws.send(Frame.Binary(true, ping(nowMs()).encode()))
            } catch (_: Exception) { active.set(false); break }
        }
    }

    // Background: TIME_NOTIFICATION every 5s
    val timeJob = launch {
        while (active.get()) {
            delay(TIME_INTERVAL_MS)
            if (!active.get()) break
            try {
                ws.send(Frame.Binary(true, timeNotif(nowMs(), nowISO()).encode()))
            } catch (_: Exception) { active.set(false); break }
        }
    }

    // Background: KISS_REQUEST every 5s
    val kissJob = launch {
        while (active.get()) {
            delay(KISS_INTERVAL_MS)
            if (!active.get()) break
            try {
                val os = System.getProperty("os.name") ?: "unknown"
                val ver = System.getProperty("os.version") ?: "unknown"
                val arch = System.getProperty("os.arch") ?: "unknown"
                ws.send(Frame.Binary(true, kissRequest(os, ver, "unknown", arch).encode()))
            } catch (_: Exception) { active.set(false); break }
        }
    }

    // Background: timeout check every 5s
    val timeoutJob = launch {
        while (active.get()) {
            delay(5000)
            if (!active.get()) break
            if (nowMs() - lastPongTs.get() > SESSION_TIMEOUT_MS) {
                log("ws-server", "[$userId] session timeout")
                active.set(false)
                break
            }
        }
    }

    // Receive loop
    try {
        for (frame in ws.incoming) {
            if (frame !is Frame.Binary) continue
            val data = frame.readBytes()

            val msg: Message
            try {
                msg = decodeMessage(data)
            } catch (e: Exception) {
                log("ws-server", "Decode error: ${e.message}")
                ws.send(Frame.Binary(true, errorMsg(ERR_DECODE, e.message ?: "decode error").encode()))
                continue
            }

            when (msg.type) {
                MSG_HELLO -> {
                    clientLanguage = msg.clientLanguage
                    log("ws-server", "HELLO from $clientLanguage, session=$sessionId, time=${nowMs()}")
                    ws.send(Frame.Binary(true, bonjour(SERVER_LANG).encode()))
                }

                MSG_ECHO_REQUEST -> {
                    log("ws-server", "ECHO_REQUEST id=${msg.echoId} meta=${msg.echoMeta} data=${msg.echoData}")
                    val resp = Message(MSG_ECHO_RESPONSE).apply {
                        echoStatus = 200
                        echoResults = listOf(EchoResult(nowMs(), 0, mapOf(
                            "id" to msg.echoId.toString(),
                            "data" to msg.echoData,
                            "meta" to msg.echoMeta,
                            "lang" to clientLanguage
                        )))
                    }
                    ws.send(Frame.Binary(true, resp.encode()))
                }

                MSG_KISS_RESPONSE -> {
                    log("ws-server", "KISS_RESPONSE lang=${msg.kissLanguage} enc=${msg.kissEncoding} tz=${msg.kissTimeZone}")
                }

                MSG_PONG -> {
                    lastPongTs.set(msg.timestampMs)
                }

                MSG_RANDOM_NUMBER -> {
                    log("ws-server", "RANDOM_NUMBER id=${msg.randomId} number=${msg.randomNumber}")
                    val hash = hashNumber(msg.randomNumber)
                    ws.send(Frame.Binary(true, hashResponseMsg(msg.randomId, hash).encode()))
                }

                MSG_DISCONNECT -> {
                    log("ws-server", "DISCONNECT reason=${msg.disconnectReason}")
                    active.set(false)
                }

                MSG_ERROR -> {
                    log("ws-server", "ERROR code=${msg.errorCode} msg=${msg.errorMessage}")
                }

                else -> {
                    val hex = "0x${msg.type.toInt().toString(16)}"
                    log("ws-server", "Unknown message type: $hex")
                    ws.send(Frame.Binary(true, errorMsg(ERR_UNKNOWN_MSG_TYPE, "unknown type $hex").encode()))
                }
            }
        }
    } catch (e: Exception) {
        log("ws-server", "[$userId] exception: ${e.message}")
    }

    // Cleanup
    active.set(false)
    pingJob.cancel()
    timeJob.cancel()
    kissJob.cancel()
    timeoutJob.cancel()
    try { ws.close(CloseReason(CloseReason.Codes.NORMAL, "server shutdown")) } catch (_: Exception) {}
    log("ws-server", "[$userId] session-")
}
